File: app/src/main/java/re/snapwi/orientation/NorrisApiController.java
package re.snapwi.orientation;

import de.greenrobot.event.EventBus;
import re.snapwi.orientation.event.JokeEvent;
import re.snapwi.orientation.io.Joke;
import re.snapwi.orientation.rest.NorrisApi;
import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class NorrisApiController {
  private NorrisApi api;
  private EventBus bus;

  public NorrisApiController(NorrisApi api, EventBus bus) {
    this.api = api;
    this.bus = bus;
    bus.register(this);
  }

  public void onEventAsync(GetRandomJokeEvent event) {
    api.randomJoke(new Callback<Joke>() {
      @Override public void success(Joke joke, Response response) {
        bus.postSticky(new JokeEvent(joke));
      }

      @Override public void failure(RetrofitError error) {
        bus.postSticky(new JokeEvent(error));
      }
    });
  }

  public static class GetRandomJokeEvent {
  }
}

File: app/src/main/kotlin/re/snapwi/orientation/rest/NorrisApiController.kt
package re.snapwi.orientation.rest

import de.greenrobot.event.EventBus
import re.snapwi.orientation.event.JokeEvent
import re.snapwi.orientation.io.Joke
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response
import javax.inject.Inject
import javax.inject.Singleton

Singleton public class NorrisApiController Inject constructor(var api: NorrisApi, var bus: EventBus) {
  init {
    bus.register(this)
  }

  public fun onEventAsync(event: GetRandomJokeEvent) {
    api.randomJoke(object : Callback<Joke> {
      override fun success(joke: Joke, response: Response) = bus.postSticky(JokeEvent(joke, null))
      override fun failure(error: RetrofitError) = bus.postSticky(JokeEvent(null, error))
    })
  }

  public class GetRandomJokeEvent
}